consensus/src/liveness/proposal_generator.rs

```diff
@@ -231,25 +230,41 @@ impl ProposalGenerator {
             // the local time exceeds it.
             let timestamp = self.time_service.get_current_timestamp();
 
-            let (max_block_txns, max_block_bytes) = if let Some(value) = chain_health_backoff {
-                let max_block_txns = self
-                    .max_block_txns
-                    .min(value.max_sending_block_txns_override);
-                let max_block_bytes = self
-                    .max_block_bytes
-                    .min(value.max_sending_block_bytes_override);
+            let mut values_max_block_txns = vec![self.max_block_txns];
+            let mut values_max_block_bytes = vec![self.max_block_bytes];
 
+            let chain_health_backoff = self
+                .chain_health_backoff_config
+                .get_backoff(round, proposer_election);
+            if let Some(value) = chain_health_backoff {
+                values_max_block_txns.push(value.max_sending_block_txns_override);
+                values_max_block_bytes.push(value.max_sending_block_bytes_override);
                 CHAIN_HEALTH_BACKOFF_TRIGGERED.observe(1);
+            } else {
+                CHAIN_HEALTH_BACKOFF_TRIGGERED.observe(0);
+            }
+
+            let consensus_backpressure = self.block_store.back_pressure();
+            if let Some(values) = if consensus_backpressure { &self.consensus_backpressure } else { &None } {
+                values_max_block_txns.push(values.max_sending_block_txns_override);
+                values_max_block_bytes.push(values.max_sending_block_bytes_override);
+                CONSENSUS_BACKPRESSURE_ON_PROPOSAL_TRIGGERED.observe(1);
+            } else {
+                CONSENSUS_BACKPRESSURE_ON_PROPOSAL_TRIGGERED.observe(0);
+            };
+
+            let max_block_txns = values_max_block_txns.into_iter().min().unwrap();
+            let max_block_bytes = values_max_block_bytes.into_iter().min().unwrap();
+
+            if consensus_backpressure || chain_health_backoff.is_some() {
```

Comment:
nit: we probably should have a separate function to calculate the backpressure, it's getting too fancy